src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/AbstractWorkspaceDocumentDiagnosticSource.cs

```diff
@@ -21,6 +26,13 @@ public static AbstractWorkspaceDocumentDiagnosticSource CreateForCodeAnalysisDia
     private sealed class FullSolutionAnalysisDiagnosticSource(TextDocument document, Func<DiagnosticAnalyzer, bool>? shouldIncludeAnalyzer)
         : AbstractWorkspaceDocumentDiagnosticSource(document)
     {
+        /// <summary>
+        /// Cached mapping between a project instance and all the diagnostics computed for it.  This is used so that
+        /// once we compute the diagnostics once for a particular project, we don't need to recompute them again as we
+        /// walk every document within it.
+        /// </summary>
+        private static readonly ConditionalWeakTable<Project, AsyncLazy<IReadOnlyList<DiagnosticData>>> s_projectToDiagnostics = new();
```

Comment:
@mavasani i didn't want to do this, but caching at the IDiagnosticAnalyzerService just seems busted currently.